### ISSUE TITLE
Exceptions only show type name and not actual useful exception information

### ISSUE THREAD

**Issue by mit-mit** [Dart-Code/Dart-Code#616]
Repro steps:

1. Enable Dart 2 preview modfe
1. Create new Flutter ap
1. Replace main.dart with the contents belowe
1. Run

=> program breaks on a _TypeError, and the debug console says: `breaking at exception: _TypeError`

Can we have an easier to understand error message?

Also, we should provide more information on what failed. If you try the same in IntelliJ or from the terminal, you get:

```
type 'String' is not a subtype of type 'int' of 'function result' where
  String is from dart:core
  int is from dart:core
```

`main.dart`:
```
import 'package:flutter/material.dart';
import 'dart:convert';

int lowestPrice() {
  List<int> prices = JSON.decode('["7", "100", "500"]');
  prices.sort();
  return prices[0];
}

void main() => runApp(new MyApp(lowestPrice()));

class MyApp extends StatelessWidget {
  final int lp;
  MyApp(this.lp);

  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      home: new Scaffold(
        appBar: new AppBar(),
        body: new Text('Lowest price $lp'),
      ),
    );
  }
}
```

**Comment by mit-mit** [comment]
cc @DanTup

**Comment by DanTup** [comment]
@devoncarew When this happens, we get an event from the debugger like this:

```js
{
	"jsonrpc": "2.0",
	"method": "streamNotify",
	"params": {
		"streamId": "Debug",
		"event": {
			"type": "Event",
			"kind": "PauseException",
			"isolate": {
				"type": "@Isolate",
				"fixedId": true,
				"id": "isolates\/491284696",
				"name": "flutter_assets:main()",
				"number": "491284696"
			},
			"timestamp": 1519632622424,
			"topFrame": {
				"type": "Frame",
				"kind": "Regular",
				"location": {
					"type": "SourceLocation",
					"script": {
						"type": "@Script",
						"fixedId": true,
						"id": "libraries\/@16291260\/scripts\/file%3A%2F%2F%2FM%3A%2FCoding%2FApplications%2Fflutter%2Fexamples%2Fhello_world%2Flib%2Fmain.dart\/161d12990ab",
						"uri": "file:\/\/\/M:\/Coding\/Applications\/flutter\/examples\/hello_world\/lib\/main.dart",
						"_kind": "kernel"
					},
					"tokenPos": 114
				},
				"function": {
					"type": "@Function",
					"fixedId": true,
					"id": "classes\/1036\/functions\/lowestPrice",
					"name": "lowestPrice",
					"owner": {
						"type": "@Library",
						"fixedId": true,
						"id": "libraries\/@16291260",
						"name": "",
						"uri": "file:\/\/\/M:\/Coding\/Applications\/flutter\/examples\/hello_world\/lib\/main.dart"
					},
					"_kind": "RegularFunction",
					"static": true,
					"const": false,
					"_intrinsic": false,
					"_native": false
				},
				"code": {
					"type": "@Code",
					"fixedId": true,
					"id": "code\/cdb0d71e4-9d1abf30",
					"name": "lowestPrice",
					"kind": "Dart",
					"_optimized": false,
					"_intrinsic": false,
					"_native": false
				},
				"vars": [],
				"index": 0
			},
			"exception": {
				"type": "@Instance",
				"class": {
					"type": "@Class",
					"fixedId": true,
					"id": "classes\/155",
					"name": "_TypeError",
					"_vmName": "_TypeError@0150898"
				},
				"kind": "PlainInstance",
				"id": "objects\/0"
			}
		}
	}
}
```

There doesn't seem to be any useful message here (and our debug implementation currently only looks at this object). Is something missing here, or should I be sending an additional request to the VM to get info on the exception?

**Comment by devoncarew** [comment]
I am surprised that the `exception` instance doesn't provide the toString() value. 

> If you try the same in IntelliJ or from the terminal, you get:

Perhaps there's some info being printed to stderr? It's odd that this is showing in the cli but not vscode.

**Comment by DanTup** [comment]
Looks like we do get it in the debug console, but only after the user continues (so at the point they break, they don't have any info):

```text
breaking at exception: _TypeError // This shows when hitting the breakpoint
// This shows if you click continue
E/flutter (25473): [ERROR:topaz/lib/tonic/logging/dart_error.cc(16)] Unhandled exception:
E/flutter (25473): type 'List' is not a subtype of type 'List<int>' where
E/flutter (25473):   List is from dart:core
E/flutter (25473):   List is from dart:core
E/flutter (25473):   int is from dart:core
E/flutter (25473):
E/flutter (25473): #0      lowestPrice (file:///M:/Coding/Applications/flutter/examples/hello_world/lib/main.dart:5:27)
E/flutter (25473): #1      main (file:///M:/Coding/Applications/flutter/examples/hello_world/lib/main.dart:10:33)
E/flutter (25473): #2      _startIsolate.<anonymous closure> (dart:isolate/isolate_patch.dart:279:19)
E/flutter (25473): #3      _RawReceivePortImpl._handleMessage (dart:isolate/isolate_patch.dart:165:12)
```

@devoncarew Do you think this might be a bug, or am I supposed to send a request back for more info about the exception?

The text "is not a subtype of" doesn't appear in my observatory log at all; only in the flutter run log.

**Comment by devoncarew** [comment]
So, I'm guessing that:
- vscode sets up the value for break on exceptions to 'uncaught'
- when Michael tried in IntelliJ above, he may have used a run configuration, not debug, so it didn't break on the exception
- running from the cli won't break on exceptions

so, vscode is breaking before the print output displays. 

> Do you think this might be a bug, or am I supposed to send a request back for more info about the exception?

I don't think there's a bug here, but it probably would be valuable for VSCode to send a round-trip to ask for the toString() of the exception and print that.

Is the exception object visible in the locals var frame? In IntelliJ, we put it in as a synthetic local - so it looks like it's a member of the top frame. I forget what's possible in the console view, but if we can add it as a first class object that people can expand - like if they'd eval'd a value from the REPL - that would be great.

**Comment by DanTup** [comment]
Aha, that makes sense now then. I guess we should be add it to the locals, but I think we need to make it more obvious than that. too.

I just started digging through the code to see how we can evaluate it, and found this:

```ts
private callToString(isolate: VMIsolateRef, instanceRef: VMInstanceRef): Promise<string> {
		return this.observatory.evaluate(isolate.id, instanceRef.id, "toString()")
```

and it's currently unused! Whoops - either it was unfinished or something got deleted! Should be a reasonable easy fix. Thanks!

**Comment by DanTup** [comment]
Oh wow, I should put a case in to review all the `TODOs`!

> // TODO: Call toString()?

```ts
if (kind === "PauseException") {
	reason = "exception";
	exceptionText = this.valueAsString(event.exception);
	if (!exceptionText)
		exceptionText = event.exception.class.name;
	// TODO: Call toString()?
	this.sendEvent(new OutputEvent(`breaking at exception: ${exceptionText}\n`));
}
```

**Comment by DanTup** [comment]
I'll open another issue about inserted into locals, but the exception info now shows the string:

![Debugger](https://user-images.githubusercontent.com/1078012/36860596-a2100dc4-1d79-11e8-8e66-353b9a98145c.png)

**Comment by DanTup** [comment]
Btw, I removed the line that was printed to the console at the time of breaking, because the unhandled exception results in the exception being dumped anyway (so you ended up with the whole exception printed twice - once at the time of breaking, and once immediately after clicking resume).

I figured that the info in the tooltip (and soon the locals window) is enough. Let me know if you're not sure (or if you want a build to play with).
